### PR TITLE
Fix CCheckQueue IsIdle (potential) race condition

### PR DESCRIPTION
**Pull request by JeremyRubin**
There's a small race condition in the CCheckQueue code which I don't think is currently an active issue, but future code might break.

IsIdle is not threadsafe. If two concurrent `CCheckqueueControl`s are made, they could simultaneously report being idle, and fail to panic.

Furthermore, in the case a concurrent `CCheckqueueControl` is made, most likely waiting until control is relinquished is the right behavior rather than failing an assert.

### CODE REVIEW

File: src/checkqueue.h [hunk 1]
```diff
@@ -185,8 +182,7 @@ class CCheckQueueControl
     {
         // passed queue is supposed to be unused, or NULL
         if (pqueue != NULL) {
-            bool isIdle = pqueue->IsIdle();
-            assert(isIdle);
+            ENTER_CRITICAL_SECTION(pqueue->ControlMutex);
```
Comment (sipa) [inline src/checkqueue.h]: Can you do this as a unique_lock field inside CCheckQueueControl (so we don't need to rely on an explicit destructor to be correct)?

Comment (JeremyRubin) [inline src/checkqueue.h]: Unfortunately this is the "easiest" way to take care of this given that we want debug order & the CMutex class has no default constructor/moveable semantics. Certainly can be fixed to RAII easily later.

Comment (sipa) [inline src/checkqueue.h]: Understood. Let's improve that in a separate PR later.

**Review by ryanofsky — APPROVED** [review]
utACK 958ee1d3d844d1f2f07b271ad7ed9bb117e5a766 with one nit:

Would suggest deleting the CCheckQueueControl default copy constructor and copy operator and making the pqueue member const, to give more assurance that the ENTER/LEAVE calls will always be paired.

File: src/checkqueue.h [hunk 2]
```diff
@@ -174,16 +174,19 @@ template <typename T>
...
     bool fDone;
 
 public:
-    CCheckQueueControl(CCheckQueue<T>* pqueueIn) : pqueue(pqueueIn), fDone(false)
-    {
-        // passed queue is supposed to be unused, or NULL
-        if (pqueue != NULL) {
-            ENTER_CRITICAL_SECTION(pqueue->ControlMutex);
-        }
+	CCheckQueueControl<T>() = delete;
```
Comment (ryanofsky) [inline src/checkqueue.h]: Nit: I think you can drop the `<T>`

**Review by ryanofsky — APPROVED** [review]
utACK 2a58f73601ad36d0531fe9e69e1affd9229d8b96

**Comment by JeremyRubin** [comment]
Squashed and addressed @ryanofsky's nit about template args in constructors.

File: src/checkqueue.h [hunk 3]
```diff
@@ -177,17 +174,19 @@ template <typename T>
...
-        if (pqueue != NULL) {
-            bool isIdle = pqueue->IsIdle();
-            assert(isIdle);
-        }
+	CCheckQueueControl() = delete;
+	CCheckQueueControl(const CCheckQueueControl&) = delete;
+	CCheckQueueControl& operator=(const CCheckQueueControl&) = delete;
+	explicit CCheckQueueControl(CCheckQueue<T> * const pqueueIn) : pqueue(pqueueIn), fDone(false)
+	{
+		// passed queue is supposed to be unused, or NULL
```
Comment (sipa) [inline src/checkqueue.h]: Weird indendation.

**Review by sipa — COMMENTED** [review]
utACK

File: src/checkqueue.h (hunk 3 above)
Comment (JeremyRubin) [inline src/checkqueue.h]: Oops -- somehow, my expandtab and tabstop=4 got unset.

**Comment by JeremyRubin** [comment]
sorry for causing extra review -- pushed and squashed indention fix.

File: src/checkqueue.h [hunk 4]
```diff
@@ -127,6 +127,9 @@ class CCheckQueue
     }
 
 public:
+    //! Mutex to ensure only one concurrent CCheckQueueControl
+    boost::mutex ControlMutex;
```
Comment (kallewoof) [inline src/checkqueue.h]: Nit: Is there some coding standard for mutexes which makes you capitalize this ivar? If not, maybe `controlMutex` would be more consistent.

**Review by kallewoof — COMMENTED** [review]
utACK f5daff7

**Comment by JeremyRubin** [comment]
#9497 was merged, which included this. Closing.

**Pull request closed without merging**